msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java

```diff
@@ -1064,51 +1064,187 @@ protected void acquireTokenSilent(@NonNull final String[] scopes,
     @Override
     public void acquireTokenSilentAsync(
             @NonNull final AcquireTokenSilentParameters acquireTokenSilentParameters) {
-        acquireTokenSilentParameters.setAccountRecord(
-                getAccountRecord(
-                        acquireTokenSilentParameters.getAccount(),
-                        getRequestTenantId(
-                                mPublicClientConfiguration,
-                                acquireTokenSilentParameters
-                        )
-                )
-        );
+        sBackgroundExecutor.submit(new Runnable() {
+            @Override
+            public void run() {
+                final ILocalAuthenticationCallback callback = getLocalAuthenticationCallback(
+                        acquireTokenSilentParameters.getCallback()
+                );
+
+                try {
+                    acquireTokenSilentParameters.setAccountRecord(
+                            selectAccountRecordForTokenRequest(
+                                    mPublicClientConfiguration,
+                                    acquireTokenSilentParameters
+                            )
+                    );
+
+                    validateAcquireTokenSilentParameters(acquireTokenSilentParameters);
+
+                    final AcquireTokenSilentOperationParameters params =
+                            OperationParametersAdapter.createAcquireTokenSilentOperationParameters(
+                                    acquireTokenSilentParameters,
+                                    mPublicClientConfiguration
+                            );
+
+
+                    final TokenCommand silentTokenCommand = new TokenCommand(
+                            params,
+                            MSALControllerFactory.getAcquireTokenSilentControllers(
+                                    mPublicClientConfiguration.getAppContext(),
+                                    params.getAuthority(),
+                                    mPublicClientConfiguration
+                            ),
+                            callback
+                    );
+
+                    ApiDispatcher.submitSilent(silentTokenCommand);
+                } catch (final BaseException exception) {
+                    callback.onError(exception);
```

Comment:
@iambmelt Why are we not posting the exception to the main thread here in silent as we do above in the interactive call?

Reply:
Because we may not be on the main thread

Reply:
What happens if we are on the main thread? The behavior I'm seeing in my tests due to this is that after callback onError method is called, the tests keeps running even after that.

Reply:
The current implementation will return the error on a background thread which in your case isn't going to work so sounds like we'll have to add the same logic here as we used in [`BrokerBaseStrategy`](https://github.com/AzureAD/microsoft-authentication-library-for-android/blob/dev/msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerBaseStrategy.java#L74-L80). I'll open a PR for it and you can test, thanks @shahzaibj

Reply:
@shahzaibj A proposed change is on `iambmelt/ats-error-thread`

Reply:
@iambmelt Thanks!